PROTOCOL.md

```diff
@@ -0,0 +1,207 @@
+  "specversion" : "1.0.1",
+  "type" : "<topic>",
+  "source" : "location or name of service",
+  "id" :  "<uuid>",
+  "time" : "2018-04-05T17:31:00Z",
+  "polyntrace": [
+    {
+      "type": "<topic>",
+      "time": "2018-04-05T17:31:00Z",
+      "id": "<uuid>"
+    }
+  ],
+  "polynclient": {
+    "lang": "ruby",
+    "langversion": "3.2.1",
+    "version": "0.1.0"
+  },
+  "datacontenttype" : "application/json",
+  "data" : {}
+  }
+}
+```
+
+j
+
+### CloudEvents Extensions
+
+Polyn Clients MUST implement the following extensions.
+
+#### `polyntrace`
+
+This is an array of trace objects that identify the chain of events that lead the current event to
+being published. A Polyn client SHOULD NOT fail to consume an event however if `polyntrace` is
+missing from the published event.
+
+##### Example
+
+```json
+[
+  {
+    "type": "<topic>",
+
+    "time": "2018-04-05T14:31:00Z",
+    "id": "<uuid>"
+  },
+  {
+    "type": "<topic>",
+    "time": "2018-04-05T17:31:00Z",
+    "id": "<uuid>"
+  }
+]
+```
+
+A Polyn client MUST provide a way to publish an event as the result of a prior event, adding the
+trace of the predecessor to the `polyntrace` array.
+
+##### Example
+
+```ruby
+def receive_some_event(context)
+  # ... do work
+  context.publish("<topic>", result_of_work)
+end
+```
+
+This would add whatever event that caused `receive_some_event` to be fired to the trace when
+`context.publish` is called.
+
+#### `polynclient`
+
+This is an object representing the information about the client that published the event. A Polyn
+client SHOULD NOT however fail to consume the event if the `polynclient` extension is missing.
+
+##### Example
+
+```json
+{
+  "lang": "ruby",
+  "langversion": "3.2.1",
+  "version": "0.1.0"
+}
+```
+
+### Additional Field Requirements
+
+#### `datacontenttype`
+
+A Polyn client SHOULD add the `datacontenttype` field as defined [here](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#datacontenttype)
+before publishing events that correctly match the content type of the `data` field. It MUST at a
+minimum be able to serialize and deserialize the `application/json` content type.
+
+If the `datacontenttype` is not present in the event, a Polyn client MUST assume that the content
+type is `application/json`. If the data cannot be deserialized, the client MUST broadcast an
+[appropriate error]().
+
+#### `data`
+
+A Polyn client MUST add its event data to the `data` attribute of the CloudEvent. The data format
+must match the `datacontentype` field.
+
+## Message Validation
+
+### JSON Schema
+
+A Polyn client MUST support loading a JSON Schema document for each event. It MUST wrap the event
+schema into a valid CloudEvent JSON Schema, and publish that schema to the Schema Repository.
+
+#### Schema Repository
+
+A Polyn client MUST publish all event schema to the Schema Repository. The repository MUST be a
+NATS KeyValue store on the same NATS server or cluster that is providing the Message Bus. The
```

Comment:
I think it would be good to specify what the name of the KeyValue Bucket should be.